packages/babel-parser/src/parser/statement.js

```diff
@@ -148,18 +144,28 @@ export default class StatementParser extends ExpressionParser {
         return this.parseDoStatement(node);
       case tt._for:
         return this.parseForStatement(node);
-      case tt._function:
+      case tt._function: {
         if (this.lookahead().type === tt.dot) break;
-        if (!declaration) {
+        if (
+          context &&
+          (this.state.strict || (context !== "if" && context !== "label"))
+        ) {
           this.raise(
             this.state.start,
             "Function declaration not allowed in this context",
           );
         }
-        return this.parseFunctionStatement(node);
+        const result = this.parseFunctionStatement(node);
+
+        // TODO: Remove this once we have proper scope tracking in place.
+        if (context && result.generator) {
+          this.unexpected(node.start);
+        }
 
+        return result;
+      }
       case tt._class:
-        if (!declaration) this.unexpected();
+        if (context) this.unexpected();
```

Comment:
As a side note, `Function declaration is not allowed in this context` is much nicer :smile: